analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -32,6 +33,40 @@ internal static class Binary
                 ? context.SetOperationConstraint(newConstraint)
                 : context.State;
 
+        public static ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper binary, bool useOpposite) =>
+            binary.OperatorKind is BinaryOperatorKind.Equals or BinaryOperatorKind.NotEquals or BinaryOperatorKind.ObjectValueEquals or BinaryOperatorKind.ObjectValueNotEquals
+                ? LearnBranchingConstraint<ObjectConstraint>(state, binary, useOpposite) ?? LearnBranchingConstraint<BoolConstraint>(state, binary, useOpposite) ?? state
+                : state;
+
+        private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBinaryOperationWrapper binary, bool useOpposite)
+            where T : SymbolicConstraint
+        {
+            if ((OperandSymbolWithoutConstraint(binary.LeftOperand) ?? OperandSymbolWithoutConstraint(binary.RightOperand)) is { } testedSymbol
+                && (OperandConstraint(binary.LeftOperand) ?? OperandConstraint(binary.RightOperand)) is { } constraint)
+            {
+                if (useOpposite ^ binary.OperatorKind is BinaryOperatorKind.NotEquals or BinaryOperatorKind.ObjectValueNotEquals)
```

Comment:
^ is quite exotic. It may help to left a short comment.
```suggestion
                if (useOpposite ^ binary.OperatorKind.IsNotEquals()) // xor
```

Reply:
This sounds like a call for a [readable language](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/operators/xor-operator) 😆 

Devs should know their language.